### DISCUSSION TITLE
What is this "microstream" about

### DISCUSSION THREAD

**Discussion by suikast42** [microstream-one/microstream#466] [General]
Hi, 

I heard about this project in devoxx 2022 talks at the first time. I am little confused about the project.

1. We are talking about object stroage here and not a database.  As I understood microstream have the capability to use many object stoages like filesystem, kafka, postgers and so on.  So the inital loading speed depends on the used stoarage right?

2. We are not talking about a database engine here right? What about stuff like ACID, Transactions, parallel data access and so on ?

**Comment by hg-ms** [comment]
Hello,

Microstream is a library that’s main usage is to fully or partially persist and restore a Java object graph, not only objects. It is different to classic databases and it not intended to replace those in all use cases.

> So the inital loading speed depends on the used stoarage right?

This is correct, the loading speed depends on the used storage target, this is also true for store operations.

> “We are not talking about a database engine here right? What about stuff like ACID, Transactions, parallel data access and so on ?”

Microstream directly accesses your applications in memory data. Those data are owned by the application, not by Microstream. In a multithreaded environment the application must ensure that no other thread modifies the (sub)graph while Microstream operates on it. E.g.: if a thread stores a java collection with Microstream no other thread is allowed to add or remove item of that collection.
This synchronization must be done by the Application.
Storing data is executed as a transaction, an atomic all-or-nothing action. When one or more entities are stored, their data is collected into a continuous block of bytes and that block is written to the physical form (the "files") in one sweep. Any problem during the IO-operation causes the whole block to be deleted (rolled back). Multiple store operations are internally isolated. 

FAQ: https://docs.microstream.one/manual/storage/faq/miscellaneous.html
